### ISSUE TITLE
Persistent nodes in multiple copies of the trees are not independent

### ISSUE THREAD

**Issue by phomer** [tendermint/merkleeyes#21]
We have a fatal bug where a persistent IAVL node that is in the current tree has been incorrectly deleted from the persistent data store. This will cause merkleeyes to crash and the tree remains invalid (does not heal itself on a restart).

The cause of the problem is the assumption that we can keep multiple copies of the tree in memory and update them independently. In specific circumstances, a change in one copy will orphan a persistent node that is required in another copy. This does not occur for the in-memory trees because Go's garbage collector correctly keeps track of the number of references. 

To trigger this problem seems to require a non-trivial tree, repetitive updates on the same key in between lots of commits. Saving historic versions delays the problem, but does not fix it.

An example of this is in tests/tendermint/orphans_test.sh for #18 

If the script [orphans.sh](https://github.com/tendermint/merkleeyes/blob/1ff30944034bdf87e26f4886c634a6125b4aa82f/tests/tendermint/orphans.sh) is run multiple times, at some point (usually within 5 runs or less) merkleeyes will crash.

If the code changes in this PR are applied, the fix will check for referenced nodes in other trees and not orphan them if they are still needed. The problem with this code is that it keeps track of external data, makes lots of assumptions about usage and probably leaks persistently as well.

**Comment by ebuchman** [comment]
I think the library needs to handle multiple instances of the tree more natively. Maybe a IAVLTree can be wrapped in a IAVLMultiTree which keeps a list of the active trees and orphans are tracked across all of them. 

The net result should be a significant semantic distinction between `tree.Copy()` and `Load(tree.Root())`

**Comment by phomer** [comment]
Yes, we could go up and just pass around a 'forest', then for all calls use an index into the correct tree. The forest would have a Save, and it would do the copies. We still need to be concerned with leaking nodes in the database, so the choice to orphan nodes would have to move out of the nodes themselves and into what is essentially a reference counter (when a node reaches 0 it is listed to be deleted from the DB).
